{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "from workflow.fastani.remove_gunc_failed_contigs_by_contamination_sp_cluster import RemoveGuncFailedContigsByContaminationSpCluster\n",
    "from workflow.config import PCT_VALUES\n",
    "from workflow.external.gtdb_metadata import GtdbMetadataR207Full, GtdbMetadataR207\n",
    "from workflow.external.gtdb_sp_clusters import GtdbSpClustersR207\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib import cm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [],
   "source": [
    "DF_SP_CLUSTERS = GtdbSpClustersR207().output().read_cached()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "data": {
      "text/plain": "                                        rep_genome  \\\nspecies                                              \ns__0-14-0-10-38-17 sp002774085  GB_GCA_002774085.1   \ns__0-14-0-20-30-16 sp002779075  GB_GCA_002779075.1   \ns__0-14-0-20-30-16 sp903916665  GB_GCA_903916665.1   \ns__0-14-0-20-34-12 sp002779065  GB_GCA_002779065.1   \ns__0-14-0-20-40-13 sp002774285  GB_GCA_002774285.1   \n\n                                                                         taxonomy  \\\nspecies                                                                             \ns__0-14-0-10-38-17 sp002774085  d__Bacteria;p__Patescibacteria;c__Microgenomat...   \ns__0-14-0-20-30-16 sp002779075  d__Archaea;p__Iainarchaeota;c__Iainarchaeia;o_...   \ns__0-14-0-20-30-16 sp903916665  d__Archaea;p__Iainarchaeota;c__Iainarchaeia;o_...   \ns__0-14-0-20-34-12 sp002779065  d__Archaea;p__Iainarchaeota;c__Iainarchaeia;o_...   \ns__0-14-0-20-40-13 sp002774285  d__Bacteria;p__Patescibacteria;c__WWE3;o__0-14...   \n\n                                ani_radius  ani_mean  ani_min  af_mean  \\\nspecies                                                                  \ns__0-14-0-10-38-17 sp002774085        95.0       NaN      NaN      NaN   \ns__0-14-0-20-30-16 sp002779075        95.0       NaN      NaN      NaN   \ns__0-14-0-20-30-16 sp903916665        95.0       NaN      NaN      NaN   \ns__0-14-0-20-34-12 sp002779065        95.0       NaN      NaN      NaN   \ns__0-14-0-20-40-13 sp002774285        95.0       NaN      NaN      NaN   \n\n                                af_min  n_genomes  \nspecies                                            \ns__0-14-0-10-38-17 sp002774085     NaN          1  \ns__0-14-0-20-30-16 sp002779075     NaN          1  \ns__0-14-0-20-30-16 sp903916665     NaN          1  \ns__0-14-0-20-34-12 sp002779065     NaN          1  \ns__0-14-0-20-40-13 sp002774285     NaN          1  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>rep_genome</th>\n      <th>taxonomy</th>\n      <th>ani_radius</th>\n      <th>ani_mean</th>\n      <th>ani_min</th>\n      <th>af_mean</th>\n      <th>af_min</th>\n      <th>n_genomes</th>\n    </tr>\n    <tr>\n      <th>species</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>s__0-14-0-10-38-17 sp002774085</th>\n      <td>GB_GCA_002774085.1</td>\n      <td>d__Bacteria;p__Patescibacteria;c__Microgenomat...</td>\n      <td>95.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>s__0-14-0-20-30-16 sp002779075</th>\n      <td>GB_GCA_002779075.1</td>\n      <td>d__Archaea;p__Iainarchaeota;c__Iainarchaeia;o_...</td>\n      <td>95.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>s__0-14-0-20-30-16 sp903916665</th>\n      <td>GB_GCA_903916665.1</td>\n      <td>d__Archaea;p__Iainarchaeota;c__Iainarchaeia;o_...</td>\n      <td>95.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>s__0-14-0-20-34-12 sp002779065</th>\n      <td>GB_GCA_002779065.1</td>\n      <td>d__Archaea;p__Iainarchaeota;c__Iainarchaeia;o_...</td>\n      <td>95.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>s__0-14-0-20-40-13 sp002774285</th>\n      <td>GB_GCA_002774285.1</td>\n      <td>d__Bacteria;p__Patescibacteria;c__WWE3;o__0-14...</td>\n      <td>95.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_SP_CLUSTERS.head()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(578539, 5)\n",
      "35,723 failed gids\n"
     ]
    },
    {
     "data": {
      "text/plain": "                          new_sp_rep      ani        af    type  same\ngid             pct                                                  \nGCA_000143435.1 1    GCF_001435955.1  97.3270  0.875591  sp_rep  True\n                5    GCF_001435955.1  97.3270  0.875591  sp_rep  True\n                15   GCF_001435955.1  97.6349  0.829228  sp_rep  True\nGCA_000153745.1 5                NaN      NaN       NaN  no_ani  True\nGCA_000155005.1 1    GCF_003697165.2  96.5203  0.827668  sp_rep  True",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th></th>\n      <th>new_sp_rep</th>\n      <th>ani</th>\n      <th>af</th>\n      <th>type</th>\n      <th>same</th>\n    </tr>\n    <tr>\n      <th>gid</th>\n      <th>pct</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th rowspan=\"3\" valign=\"top\">GCA_000143435.1</th>\n      <th>1</th>\n      <td>GCF_001435955.1</td>\n      <td>97.3270</td>\n      <td>0.875591</td>\n      <td>sp_rep</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>GCF_001435955.1</td>\n      <td>97.3270</td>\n      <td>0.875591</td>\n      <td>sp_rep</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>GCF_001435955.1</td>\n      <td>97.6349</td>\n      <td>0.829228</td>\n      <td>sp_rep</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>GCA_000153745.1</th>\n      <th>5</th>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>no_ani</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>GCA_000155005.1</th>\n      <th>1</th>\n      <td>GCF_003697165.2</td>\n      <td>96.5203</td>\n      <td>0.827668</td>\n      <td>sp_rep</td>\n      <td>True</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF = RemoveGuncFailedContigsByContaminationSpCluster().output().read_cached()\n",
    "print(DF.shape)\n",
    "\n",
    "UNQ_GIDS = set(DF.index.get_level_values(0))\n",
    "\n",
    "print(f'{len(UNQ_GIDS):,} failed gids')\n",
    "DF.head()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:2528\u001B[0m, in \u001B[0;36mDataCol.convert\u001B[0;34m(self, values, nan_rep, encoding, errors)\u001B[0m\n\u001B[1;32m   2527\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m-> 2528\u001B[0m     converted \u001B[38;5;241m=\u001B[39m \u001B[43mconverted\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mastype\u001B[49m\u001B[43m(\u001B[49m\u001B[43mdtype\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mcopy\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43;01mFalse\u001B[39;49;00m\u001B[43m)\u001B[49m\n\u001B[1;32m   2529\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mTypeError\u001B[39;00m:\n",
      "\u001B[0;31mTypeError\u001B[0m: data type 'bytes5952' not understood",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "Input \u001B[0;32mIn [5]\u001B[0m, in \u001B[0;36m<cell line: 1>\u001B[0;34m()\u001B[0m\n\u001B[0;32m----> 1\u001B[0m DF_META \u001B[38;5;241m=\u001B[39m \u001B[43mGtdbMetadataR207Full\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43moutput\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread_cached\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m      2\u001B[0m DF_META\u001B[38;5;241m.\u001B[39mhead()\n",
      "File \u001B[0;32m~/git/phd-gunc-chimeras/workflow/model/luigi.py:154\u001B[0m, in \u001B[0;36mLocalTargetHdf5.read_cached\u001B[0;34m(self, key, where, start, stop, columns, iterator, chunksize)\u001B[0m\n\u001B[1;32m    146\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mread_cached\u001B[39m(\u001B[38;5;28mself\u001B[39m, key: Optional[\u001B[38;5;28mstr\u001B[39m] \u001B[38;5;241m=\u001B[39m \u001B[38;5;124m'\u001B[39m\u001B[38;5;124mroot\u001B[39m\u001B[38;5;124m'\u001B[39m,\n\u001B[1;32m    147\u001B[0m                 where: Optional[\u001B[38;5;28mstr\u001B[39m] \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m    148\u001B[0m                 start: Optional[\u001B[38;5;28mint\u001B[39m] \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m    151\u001B[0m                 iterator: Optional[\u001B[38;5;28mbool\u001B[39m] \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m    152\u001B[0m                 chunksize: Optional[\u001B[38;5;28mint\u001B[39m] \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m) \u001B[38;5;241m-\u001B[39m\u001B[38;5;241m>\u001B[39m pd\u001B[38;5;241m.\u001B[39mDataFrame:\n\u001B[1;32m    153\u001B[0m     path \u001B[38;5;241m=\u001B[39m maybe_cache(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mpath)\n\u001B[0;32m--> 154\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mpd\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread_hdf\u001B[49m\u001B[43m(\u001B[49m\u001B[43mpath\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mkey\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mkey\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mwhere\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mwhere\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstart\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstart\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    155\u001B[0m \u001B[43m                       \u001B[49m\u001B[43mstop\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstop\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mcolumns\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mcolumns\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43miterator\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43miterator\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    156\u001B[0m \u001B[43m                       \u001B[49m\u001B[43mchunksize\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mchunksize\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:455\u001B[0m, in \u001B[0;36mread_hdf\u001B[0;34m(path_or_buf, key, mode, errors, where, start, stop, columns, iterator, chunksize, **kwargs)\u001B[0m\n\u001B[1;32m    450\u001B[0m                 \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\n\u001B[1;32m    451\u001B[0m                     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mkey must be provided when HDF5 \u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    452\u001B[0m                     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mfile contains multiple datasets.\u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    453\u001B[0m                 )\n\u001B[1;32m    454\u001B[0m         key \u001B[38;5;241m=\u001B[39m candidate_only_group\u001B[38;5;241m.\u001B[39m_v_pathname\n\u001B[0;32m--> 455\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mstore\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mselect\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m    456\u001B[0m \u001B[43m        \u001B[49m\u001B[43mkey\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    457\u001B[0m \u001B[43m        \u001B[49m\u001B[43mwhere\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mwhere\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    458\u001B[0m \u001B[43m        \u001B[49m\u001B[43mstart\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstart\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    459\u001B[0m \u001B[43m        \u001B[49m\u001B[43mstop\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstop\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    460\u001B[0m \u001B[43m        \u001B[49m\u001B[43mcolumns\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mcolumns\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    461\u001B[0m \u001B[43m        \u001B[49m\u001B[43miterator\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43miterator\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    462\u001B[0m \u001B[43m        \u001B[49m\u001B[43mchunksize\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mchunksize\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    463\u001B[0m \u001B[43m        \u001B[49m\u001B[43mauto_close\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mauto_close\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    464\u001B[0m \u001B[43m    \u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    465\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m (\u001B[38;5;167;01mValueError\u001B[39;00m, \u001B[38;5;167;01mTypeError\u001B[39;00m, \u001B[38;5;167;01mKeyError\u001B[39;00m):\n\u001B[1;32m    466\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(path_or_buf, HDFStore):\n\u001B[1;32m    467\u001B[0m         \u001B[38;5;66;03m# if there is an error, close the store if we opened it.\u001B[39;00m\n",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:875\u001B[0m, in \u001B[0;36mHDFStore.select\u001B[0;34m(self, key, where, start, stop, columns, iterator, chunksize, auto_close)\u001B[0m\n\u001B[1;32m    861\u001B[0m \u001B[38;5;66;03m# create the iterator\u001B[39;00m\n\u001B[1;32m    862\u001B[0m it \u001B[38;5;241m=\u001B[39m TableIterator(\n\u001B[1;32m    863\u001B[0m     \u001B[38;5;28mself\u001B[39m,\n\u001B[1;32m    864\u001B[0m     s,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m    872\u001B[0m     auto_close\u001B[38;5;241m=\u001B[39mauto_close,\n\u001B[1;32m    873\u001B[0m )\n\u001B[0;32m--> 875\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mit\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_result\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:1943\u001B[0m, in \u001B[0;36mTableIterator.get_result\u001B[0;34m(self, coordinates)\u001B[0m\n\u001B[1;32m   1940\u001B[0m     where \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mwhere\n\u001B[1;32m   1942\u001B[0m \u001B[38;5;66;03m# directly return the result\u001B[39;00m\n\u001B[0;32m-> 1943\u001B[0m results \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mfunc\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mstart\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mstop\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mwhere\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   1944\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mclose()\n\u001B[1;32m   1945\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m results\n",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:859\u001B[0m, in \u001B[0;36mHDFStore.select.<locals>.func\u001B[0;34m(_start, _stop, _where)\u001B[0m\n\u001B[1;32m    858\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mfunc\u001B[39m(_start, _stop, _where):\n\u001B[0;32m--> 859\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43ms\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread\u001B[49m\u001B[43m(\u001B[49m\u001B[43mstart\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m_start\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstop\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m_stop\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mwhere\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m_where\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mcolumns\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mcolumns\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:4531\u001B[0m, in \u001B[0;36mAppendableFrameTable.read\u001B[0;34m(self, where, columns, start, stop)\u001B[0m\n\u001B[1;32m   4528\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39minfer_axes():\n\u001B[1;32m   4529\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[0;32m-> 4531\u001B[0m result \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_read_axes\u001B[49m\u001B[43m(\u001B[49m\u001B[43mwhere\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mwhere\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstart\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstart\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstop\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstop\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   4533\u001B[0m info \u001B[38;5;241m=\u001B[39m (\n\u001B[1;32m   4534\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39minfo\u001B[38;5;241m.\u001B[39mget(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mnon_index_axes[\u001B[38;5;241m0\u001B[39m][\u001B[38;5;241m0\u001B[39m], {})\n\u001B[1;32m   4535\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mlen\u001B[39m(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mnon_index_axes)\n\u001B[1;32m   4536\u001B[0m     \u001B[38;5;28;01melse\u001B[39;00m {}\n\u001B[1;32m   4537\u001B[0m )\n\u001B[1;32m   4539\u001B[0m inds \u001B[38;5;241m=\u001B[39m [i \u001B[38;5;28;01mfor\u001B[39;00m i, ax \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28menumerate\u001B[39m(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39maxes) \u001B[38;5;28;01mif\u001B[39;00m ax \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mindex_axes[\u001B[38;5;241m0\u001B[39m]]\n",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:3729\u001B[0m, in \u001B[0;36mTable._read_axes\u001B[0;34m(self, where, start, stop)\u001B[0m\n\u001B[1;32m   3727\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m a \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39maxes:\n\u001B[1;32m   3728\u001B[0m     a\u001B[38;5;241m.\u001B[39mset_info(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39minfo)\n\u001B[0;32m-> 3729\u001B[0m     res \u001B[38;5;241m=\u001B[39m \u001B[43ma\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mconvert\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m   3730\u001B[0m \u001B[43m        \u001B[49m\u001B[43mvalues\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   3731\u001B[0m \u001B[43m        \u001B[49m\u001B[43mnan_rep\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mnan_rep\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   3732\u001B[0m \u001B[43m        \u001B[49m\u001B[43mencoding\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mencoding\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   3733\u001B[0m \u001B[43m        \u001B[49m\u001B[43merrors\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43merrors\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   3734\u001B[0m \u001B[43m    \u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   3735\u001B[0m     results\u001B[38;5;241m.\u001B[39mappend(res)\n\u001B[1;32m   3737\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m results\n",
      "File \u001B[0;32m~/opt/miniconda3/envs/gunc-chimeras/lib/python3.8/site-packages/pandas/io/pytables.py:2530\u001B[0m, in \u001B[0;36mDataCol.convert\u001B[0;34m(self, values, nan_rep, encoding, errors)\u001B[0m\n\u001B[1;32m   2528\u001B[0m         converted \u001B[38;5;241m=\u001B[39m converted\u001B[38;5;241m.\u001B[39mastype(dtype, copy\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mFalse\u001B[39;00m)\n\u001B[1;32m   2529\u001B[0m     \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mTypeError\u001B[39;00m:\n\u001B[0;32m-> 2530\u001B[0m         converted \u001B[38;5;241m=\u001B[39m \u001B[43mconverted\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mastype\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mO\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mcopy\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43;01mFalse\u001B[39;49;00m\u001B[43m)\u001B[49m\n\u001B[1;32m   2532\u001B[0m \u001B[38;5;66;03m# convert nans / decode\u001B[39;00m\n\u001B[1;32m   2533\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m _ensure_decoded(kind) \u001B[38;5;241m==\u001B[39m \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mstring\u001B[39m\u001B[38;5;124m\"\u001B[39m:\n",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "source": [
    "DF_META = GtdbMetadataR207Full().output().read_cached()\n",
    "DF_META.head()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "types:\n",
    "no_af = no alignment fraction <0.5\n",
    "no_ani = novel species cluster\n",
    "sp_rep = found species representative within ANI radius\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from workflow.util.paths import get_gid_root\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "\n",
    "PCT_VALUES = [0, 1, 5, 10, 15, 20, 30, 40, 50]\n",
    "\n",
    "cmap  = sns.color_palette('bright', n_colors=len(PCT_VALUES), as_cmap=False)\n",
    "\n",
    "\n",
    "def get_data_for_gid(gid):\n",
    "    gid_root = get_gid_root(gid)\n",
    "\n",
    "    try:\n",
    "        base_ani = pd.read_hdf(os.path.join(gid_root, 'fastani_gunc_failed_pct_0.h5'))\n",
    "        pct_ani = pd.read_hdf(os.path.join(gid_root, 'fastani_gunc_failed_by_contamination.h5'))\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print(f\"missing: {gid}\")\n",
    "        return\n",
    "\n",
    "    df = pd.concat([base_ani, pct_ani], ignore_index=True)\n",
    "\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "MIN_ANI = 93\n",
    "\n",
    "\n",
    "def xform(x):\n",
    "    return 1.5 * np.pi * (x-MIN_ANI)/(100-MIN_ANI)\n",
    "\n",
    "\n",
    "def do_single_plot(gid, df):\n",
    "\n",
    "    df_old = df.copy()\n",
    "\n",
    "    meta_row = DF_META.loc[gid]\n",
    "    cur_sp_radius = DF_SP_CLUSTERS.loc[meta_row['species'], 'ani_radius']\n",
    "\n",
    "    d_rep_to_values = defaultdict(list)\n",
    "    d_rep_to_pct = defaultdict(list)\n",
    "\n",
    "    is_rep = meta_row['gtdb_genome_representative'][3:] == gid\n",
    "\n",
    "    df = df[df['reference'] != gid]\n",
    "\n",
    "    df = df[df['pct'].isin(PCT_VALUES)]\n",
    "\n",
    "\n",
    "    df = df[(df['af'] >= 0.5) & (df['ani'] >= MIN_ANI)]\n",
    "\n",
    "    if len(df) == 0:\n",
    "        print(f'no info left for: {gid}')\n",
    "        return\n",
    "\n",
    "    df = df.sort_values(by=['reference', 'pct'])\n",
    "\n",
    "    reps_that_exceed_radius = frozenset(df[df['ani'] >= cur_sp_radius]['reference'])\n",
    "\n",
    "    pct_to_rep_point = dict()\n",
    "    pct_to_rep_point_correct = dict()\n",
    "    for _, row in df.iterrows():\n",
    "        cur_ani = row['ani']\n",
    "        cur_pct = row['pct']\n",
    "        d_rep_to_values[row['reference']].append(cur_ani)\n",
    "        d_rep_to_pct[row['reference']].append(cur_pct)\n",
    "\n",
    "        if cur_ani >= cur_sp_radius:\n",
    "            prev_pct_to_rep_point = pct_to_rep_point.get(cur_pct, 0)\n",
    "            if cur_ani > prev_pct_to_rep_point:\n",
    "                pct_to_rep_point_correct[cur_pct] = DF_META.loc[row['reference'], 'species'] == meta_row['species']\n",
    "                pct_to_rep_point[cur_pct] = cur_ani\n",
    "\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(10, 10), subplot_kw={'projection': 'polar'})\n",
    "    plt.rcParams.update({'font.size': 12})\n",
    "    plt.rcParams['svg.fonttype'] = 'none'\n",
    "    next_colour_i = 0\n",
    "    for i, (rep, ani) in enumerate(d_rep_to_values.items()):\n",
    "        pct = d_rep_to_pct[rep]\n",
    "\n",
    "\n",
    "        # ax.plot([0, np.pi/5, np.pi/2, np.pi], [0, 10, 20, 30] )\n",
    "        #\n",
    "        # cur_ani = min_max(ani, 0, np.pi * 1.5)\n",
    "\n",
    "\n",
    "        cur_ani = [xform(x) for x in ani]\n",
    "\n",
    "        if rep in reps_that_exceed_radius:\n",
    "            if DF_META.loc[rep, 'species'] == meta_row['species']:\n",
    "                marker = 'o'\n",
    "            else:\n",
    "                marker = 'D'\n",
    "            ax.plot(cur_ani, pct, '-', marker=marker, label=DF_META.loc[rep, 'species'],\n",
    "                    color=cmap[next_colour_i % len(cmap)], zorder=20)\n",
    "            next_colour_i += 1\n",
    "        else:\n",
    "            ax.plot(cur_ani, pct, '-', marker='o', alpha=0.1, color='gray')\n",
    "\n",
    "        # break\n",
    "        #\n",
    "        # if i > 0:\n",
    "        #     break\n",
    "\n",
    "\n",
    "\n",
    "        # for cur_ani, cur_pct in zip(ani, pct):\n",
    "        #     print(cur_ani, cur_pct)\n",
    "        #     rep_sp = DF_META.loc[rep, 'species']\n",
    "        #     ax.scatter(cur_ani, rep_sp, color=cmap[PCT_VALUES.index(cur_pct)], alpha=0.4)\n",
    "\n",
    "    # ax.set_xlim([X_LIM[0] - 1, X_LIM[1] + 1])\n",
    "\n",
    "    # Put a special marker at each representative point\n",
    "    for cur_pct, cur_ani in pct_to_rep_point.items():\n",
    "        is_correct = pct_to_rep_point_correct[cur_pct]\n",
    "        if is_correct:\n",
    "            ax.scatter(xform(cur_ani), cur_pct, color='g', marker='x', zorder=100, alpha=1, s=120)\n",
    "        else:\n",
    "            ax.scatter(xform(cur_ani), cur_pct, color='r', marker='x', zorder=100, alpha=1, s=120)\n",
    "\n",
    "    ax.vlines(x=xform(cur_sp_radius), ymin=-10, ymax=70, color='r',\n",
    "              linestyle='--', alpha=0.4, label=f'ANI Radius: {cur_sp_radius:.2f}%')\n",
    "\n",
    "    new_xticks, new_xlabels = list(), list()\n",
    "    for i in range(MIN_ANI, 101):\n",
    "        new_xticks.append(xform(i))\n",
    "        new_xlabels.append(str(i))\n",
    "\n",
    "    new_yticks, new_ylabels = list(), list()\n",
    "    for i in [-10] + PCT_VALUES:\n",
    "        new_yticks.append(i)\n",
    "        new_ylabels.append(str(i) if i not in {-10, 0} else '')\n",
    "\n",
    "    ax.grid(True)\n",
    "    ax.set_xlim([0, np.pi * 1.5])\n",
    "    ax.set_ylim([-10, max(PCT_VALUES) + 5])\n",
    "    ax.set_theta_direction(-1)\n",
    "\n",
    "    ax.set_xticklabels(new_xlabels)\n",
    "    ax.set_xticks(new_xticks)\n",
    "    # ax.set_xlabel('% of genome removed')\n",
    "\n",
    "    ax.set_yticks(new_yticks)\n",
    "    ax.set_yticklabels(new_ylabels)\n",
    "    ax.set_ylabel('% ANI')\n",
    "\n",
    "    ax.text(x=-0.35, y=12, s='% of genome removed')\n",
    "\n",
    "    if is_rep:\n",
    "        title_text = f'{gid} (species representative)\\n{meta_row[\"species\"]}'\n",
    "    else:\n",
    "        title_text = f'{gid}\\n{meta_row[\"species\"]}'\n",
    "\n",
    "    plt.title(title_text)\n",
    "\n",
    "    plt.legend()\n",
    "\n",
    "\n",
    "\n",
    "    # plt.show()\n",
    "\n",
    "    os.makedirs('/tmp/guncplots', exist_ok=True)\n",
    "    plt.savefig(f\"/tmp/guncplots/{gid}_rep_{is_rep}.svg\")\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "    print(gid)\n",
    "    print('^^^^^^')\n",
    "    return\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def gen_data():\n",
    "\n",
    "    # Some were originally run with more pct values, but just reduce to the analysis set\n",
    "    df_changed = DF[DF['same'] == False]\n",
    "    df_changed = df_changed[df_changed.index.get_level_values(1).isin(PCT_VALUES)]\n",
    "\n",
    "    d_pct_to_changes = defaultdict(list)\n",
    "\n",
    "    gids_to_check = frozenset(df_changed.index.get_level_values(0))\n",
    "    print(f'{len(gids_to_check):,} gids to check')\n",
    "\n",
    "    GIDS_TO_KEEP ={'GCF_015643835.1', 'GCA_015257755.1', 'GCF_900509435.1', 'GCA_013213925.1', 'GCA_902560595.1', 'GCA_017852475.1', 'GCA_900761595.1', 'GCA_018239885.1', 'GCA_900759145.1', 'GCA_903931905.1', 'GCA_017515185.1', 'GCA_008668795.1', 'GCA_018056875.1', 'GCA_902593295.1', 'GCA_018662785.1', 'GCA_002731855.1', 'GCA_009619015.1', 'GCF_000698005.1', 'GCA_017394825.1', 'GCA_902528895.1', 'GCA_900763945.1', 'GCA_008668585.1', 'GCA_900759525.1', 'GCA_903846615.1', 'GCA_002722235.1', 'GCA_905201125.1', 'GCA_001509115.1', 'GCA_900760075.1', 'GCA_905214645.1', 'GCA_905200745.1', 'GCA_017465765.1', 'GCA_905208535.1', 'GCA_900555595.1', 'GCA_009493725.1', 'GCA_900761055.1', 'GCA_011523145.1', 'GCA_900765305.1', 'GCF_002883995.1'}\n",
    "\n",
    "    for gid in gids_to_check:\n",
    "\n",
    "        # if gid not in GIDS_TO_KEEP:\n",
    "        #     continue\n",
    "\n",
    "        gid_df = get_data_for_gid(gid)\n",
    "\n",
    "        if gid_df is None:\n",
    "            continue\n",
    "\n",
    "        do_single_plot(gid, gid_df)\n",
    "\n",
    "        # break\n",
    "\n",
    "\n",
    "# data = gen_data()\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from workflow.util.paths import get_gid_root\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "\n",
    "PCT_VALUES = [0, 1, 5, 10, 15, 20, 30, 40, 50]\n",
    "\n",
    "cmap  = sns.color_palette('bright', n_colors=len(PCT_VALUES), as_cmap=False)\n",
    "\n",
    "\n",
    "def get_data_for_gid(gid):\n",
    "    gid_root = get_gid_root(gid)\n",
    "\n",
    "    try:\n",
    "        base_ani = pd.read_hdf(os.path.join(gid_root, 'fastani_gunc_failed_pct_0.h5'))\n",
    "        pct_ani = pd.read_hdf(os.path.join(gid_root, 'fastani_gunc_failed_by_contamination.h5'))\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print(f\"missing: {gid}\")\n",
    "        return\n",
    "\n",
    "    df = pd.concat([base_ani, pct_ani], ignore_index=True)\n",
    "\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "MIN_ANI = 93\n",
    "\n",
    "\n",
    "def xform(x):\n",
    "    return 1.5 * np.pi * (x-MIN_ANI)/(100-MIN_ANI)\n",
    "\n",
    "\n",
    "def do_single_plot(gid, df):\n",
    "\n",
    "    df_old = df.copy()\n",
    "\n",
    "    meta_row = DF_META.loc[gid]\n",
    "    cur_sp_radius = DF_SP_CLUSTERS.loc[meta_row['species'], 'ani_radius']\n",
    "\n",
    "    d_rep_to_values = defaultdict(list)\n",
    "    d_rep_to_pct = defaultdict(list)\n",
    "\n",
    "    is_rep = meta_row['gtdb_genome_representative'][3:] == gid\n",
    "\n",
    "    df = df[df['reference'] != gid]\n",
    "\n",
    "    df = df[df['pct'].isin(PCT_VALUES)]\n",
    "\n",
    "\n",
    "    df = df[(df['af'] >= 0.5) & (df['ani'] >= MIN_ANI)]\n",
    "\n",
    "    if len(df) == 0:\n",
    "        print(f'no info left for: {gid}')\n",
    "        return\n",
    "\n",
    "    df = df.sort_values(by=['reference', 'pct'])\n",
    "\n",
    "    reps_that_exceed_radius = frozenset(df[df['ani'] >= cur_sp_radius]['reference'])\n",
    "\n",
    "    pct_to_rep_point = dict()\n",
    "    pct_to_rep_point_correct = dict()\n",
    "    for _, row in df.iterrows():\n",
    "        cur_ani = row['ani']\n",
    "        cur_pct = row['pct']\n",
    "        d_rep_to_values[row['reference']].append(cur_ani)\n",
    "        d_rep_to_pct[row['reference']].append(cur_pct)\n",
    "\n",
    "        if cur_ani >= cur_sp_radius:\n",
    "            prev_pct_to_rep_point = pct_to_rep_point.get(cur_pct, 0)\n",
    "            if cur_ani > prev_pct_to_rep_point:\n",
    "                pct_to_rep_point_correct[cur_pct] = DF_META.loc[row['reference'], 'species'] == meta_row['species']\n",
    "                pct_to_rep_point[cur_pct] = cur_ani\n",
    "\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(10, 10))\n",
    "    plt.rcParams.update({'font.size': 12})\n",
    "\n",
    "    plt.rcParams['svg.fonttype'] = 'none'\n",
    "    next_colour_i = 0\n",
    "    for i, (rep, ani) in enumerate(d_rep_to_values.items()):\n",
    "        pct = d_rep_to_pct[rep]\n",
    "\n",
    "\n",
    "        # ax.plot([0, np.pi/5, np.pi/2, np.pi], [0, 10, 20, 30] )\n",
    "        #\n",
    "        # cur_ani = min_max(ani, 0, np.pi * 1.5)\n",
    "\n",
    "        if rep in reps_that_exceed_radius:\n",
    "            if DF_META.loc[rep, 'species'] == meta_row['species']:\n",
    "                marker = 'o'\n",
    "            else:\n",
    "                marker = 'D'\n",
    "            ax.plot(pct, ani, '-', marker=marker, label=DF_META.loc[rep, 'species'],\n",
    "                    color=cmap[next_colour_i % len(cmap)], zorder=20)\n",
    "            next_colour_i += 1\n",
    "\n",
    "\n",
    "            print(pct)\n",
    "            print(ani)\n",
    "            print(f'----------^^^  {DF_META.loc[rep, \"species\"]} ^^^-----------')\n",
    "\n",
    "\n",
    "        else:\n",
    "            ax.plot(pct, ani, '-', marker='o', alpha=0.1, color='gray')\n",
    "\n",
    "        # break\n",
    "        #\n",
    "        # if i > 0:\n",
    "        #     break\n",
    "\n",
    "\n",
    "\n",
    "        # for cur_ani, cur_pct in zip(ani, pct):\n",
    "        #     print(cur_ani, cur_pct)\n",
    "        #     rep_sp = DF_META.loc[rep, 'species']\n",
    "        #     ax.scatter(cur_ani, rep_sp, color=cmap[PCT_VALUES.index(cur_pct)], alpha=0.4)\n",
    "\n",
    "    # ax.set_xlim([X_LIM[0] - 1, X_LIM[1] + 1])\n",
    "\n",
    "    # Put a special marker at each representative point\n",
    "    for cur_pct, cur_ani in pct_to_rep_point.items():\n",
    "        is_correct = pct_to_rep_point_correct[cur_pct]\n",
    "        if is_correct:\n",
    "            ax.scatter(cur_pct,cur_ani , color='g', marker='o', zorder=1, s=120)\n",
    "        else:\n",
    "            ax.scatter(cur_pct, cur_ani, color='r', marker='o', zorder=1,  s=120)\n",
    "\n",
    "    ax.hlines(y=cur_sp_radius, xmin=-10, xmax=70, color='r',\n",
    "              linestyle='--', alpha=0.4, label=f'ANI Radius: {cur_sp_radius:.2f}%')\n",
    "\n",
    "    new_xticks, new_xlabels = list(), list()\n",
    "    for i in PCT_VALUES:\n",
    "        new_xticks.append(i)\n",
    "        new_xlabels.append(str(i))\n",
    "\n",
    "    new_yticks, new_ylabels = list(), list()\n",
    "    for i in [-10] + PCT_VALUES:\n",
    "        new_yticks.append(i)\n",
    "        new_ylabels.append(str(i) if i not in {-10, 0} else '')\n",
    "\n",
    "    ax.grid(True)\n",
    "    ax.set_xlim(-1, 51)\n",
    "    # ax.set_ylim([-10, max(PCT_VALUES) + 5])\n",
    "\n",
    "    ax.set_ylim([94, 100])\n",
    "\n",
    "    ax.set_xticklabels(new_xlabels)\n",
    "    ax.set_xticks(new_xticks)\n",
    "    ax.set_xlabel('% of genome removed')\n",
    "\n",
    "    # ax.set_yticks(new_yticks)\n",
    "    # ax.set_yticklabels(new_ylabels)\n",
    "    ax.set_ylabel('% ANI')\n",
    "\n",
    "    # ax.text(x=-0.35, y=12, s='% of genome removed')\n",
    "\n",
    "    if is_rep:\n",
    "        title_text = f'{gid} (species representative)\\n{meta_row[\"species\"]}'\n",
    "    else:\n",
    "        title_text = f'{gid}\\n{meta_row[\"species\"]}'\n",
    "\n",
    "    plt.title(title_text)\n",
    "\n",
    "    plt.legend()\n",
    "\n",
    "    ncbi_cat = DF_META.loc[gid, 'ncbi_genome_category']\n",
    "\n",
    "    # plt.show()\n",
    "\n",
    "\n",
    "\n",
    "    os.makedirs('/tmp/guncplots', exist_ok=True)\n",
    "    plt.savefig(f\"/tmp/guncplots/{gid}_rep_{is_rep}_{ncbi_cat}.svg\")\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "    print(gid)\n",
    "    print(ncbi_cat)\n",
    "    print('^^^^^^')\n",
    "    return\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def gen_data():\n",
    "\n",
    "    # Some were originally run with more pct values, but just reduce to the analysis set\n",
    "    df_changed = DF[DF['same'] == False]\n",
    "    df_changed = df_changed[df_changed.index.get_level_values(1).isin(PCT_VALUES)]\n",
    "\n",
    "\n",
    "\n",
    "    # df_changed = df_changed[df_changed.index.get_level_values(1).isin({0, 1, 5, 10})]\n",
    "\n",
    "    d_pct_to_changes = defaultdict(list)\n",
    "\n",
    "    gids_to_check = frozenset(df_changed.index.get_level_values(0))\n",
    "    print(f'{len(gids_to_check):,} gids to check')\n",
    "\n",
    "    GIDS_TO_KEEP ={'GCF_002026585.1', 'GCA_900751995.1', 'GCA_900759445.1'}\n",
    "\n",
    "    for gid in gids_to_check:\n",
    "\n",
    "        if gid not in GIDS_TO_KEEP:\n",
    "            continue\n",
    "\n",
    "        gid_df = get_data_for_gid(gid)\n",
    "\n",
    "\n",
    "        do_single_plot(gid, gid_df)\n",
    "\n",
    "\n",
    "        # break\n",
    "\n",
    "\n",
    "data = gen_data()\n",
    "# DF[DF['same'] == False]"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}