(window.webpackJsonp=window.webpackJsonp||[]).push([[7],{311:function(t,e,n){t.exports=n.p+"img/ace_logo.826bc9e.png"},312:function(t,e,n){t.exports=n.p+"img/contact.94693e8.svg"},313:function(t,e,n){t.exports=n.p+"img/uq_logo.2e745a6.svg"},332:function(t,e,n){var content=n(365);content.__esModule&&(content=content.default),"string"==typeof content&&(content=[[t.i,content,""]]),content.locals&&(t.exports=content.locals);(0,n(61).default)("53620f1c",content,!0,{sourceMap:!1})},351:function(t,e,n){t.exports=n.p+"img/methods.ddbf1a1.svg"},352:function(t,e,n){t.exports=n.p+"img/sankey.67d7c54.svg"},353:function(t,e,n){t.exports=n.p+"img/contig_removal.ce8e326.svg"},354:function(t,e,n){t.exports=n.p+"img/kleb.29259da.svg"},355:function(t,e,n){t.exports=n.p+"img/wga.7bf203f.svg"},356:function(t,e,n){t.exports=n.p+"img/kleb_ani.34ac724.svg"},357:function(t,e,n){t.exports=n.p+"img/helicobacter.64b305a.svg"},358:function(t,e,n){t.exports=n.p+"img/ani.4b1c4c0.svg"},364:function(t,e,n){"use strict";n(332)},365:function(t,e,n){var c=n(60)(!1);c.push([t.i,".text-white[data-v-0b2808f0]{color:#fff!important}.bg-purple[data-v-0b2808f0]{background-color:#4f2984!important}",""]),t.exports=c},381:function(t,e,n){"use strict";n.r(e);var c=n(336),o=n(309),r=n(379),l=n(378),m={name:"IndexPage"},h=(n(364),n(75)),component=Object(h.a)(m,(function(){var t=this,e=t._self._c;return e("div",[e(c.a,{staticClass:"mt-3",attrs:{color:"#4f2984"}},[e(o.c,{staticClass:"justify-center text-white"},[t._v("\n      Genome contamination has minimal impact on the delineation of species\n    ")]),t._v(" "),e(o.a,{staticClass:"text-center text-white"},[t._v("\n      Aaron J. Mussig, Pierre-Alain Chaumeil, Maria Chuvochina, Christian Rinke, Donovan H. Parks, Philip\n      Hugenholtz"),e("br"),t._v("\n      The University of Queensland, School of Chemistry and Molecular Biosciences, Australian Centre for Ecogenomics,\n      St Lucia, QLD, Australia\n    ")])],1),t._v(" "),e(c.a,{staticClass:"mt-3"},[e(o.b,{staticClass:"text-h6 text-center"},[t._v("\n      Contamination is a known issue, but to what extent does this impact taxonomy?"),e("br"),t._v("\n      We assess the impact of putative contamination on species delineated using average nucleotide identity (ANI).\n    ")])],1),t._v(" "),e(c.a,{staticClass:"mt-3"},[e(o.c,{staticClass:"justify-center text-white bg-purple"},[t._v("\n      Method\n    ")]),t._v(" "),e(o.b,{staticClass:"text-center pt-4"},[e("img",{staticStyle:{"max-height":"400px"},attrs:{alt:"methods",src:n(351)}})])],1),t._v(" "),e(c.a,{staticClass:"mt-3"},[e(o.c,{staticClass:"justify-center text-white bg-purple"},[t._v("\n      Species clustering is robust to contamination\n    ")]),t._v(" "),e(o.b,{staticClass:"text-center pt-4"},[e(c.a,{attrs:{outlined:""}},[e(o.b,[e("img",{staticStyle:{"max-height":"300px"},attrs:{alt:"methods",src:n(352)}}),t._v(" "),e("div",{staticClass:"mt-2"},[t._v("\n            • 35,723 of 317,542 (11.25%) GTDB R207 genomes were putatively contaminated."),e("br"),t._v("\n            • Identifying contamination using a reference database has limitations:"),e("br"),t._v("\n            i) Taxonomically novel genomes are insufficiently represented."),e("br"),t._v("\n            ii) Order that genomes appear in the GUNC reference database impacts results.\n          ")])])],1),t._v(" "),e(c.a,{staticClass:"mt-4",attrs:{outlined:""}},[e(o.c,{staticClass:"justify-center"},[t._v("\n          Up to 226 genomes (0.07%) change species assignment when removing putatively contaminated contigs\n        ")]),t._v(" "),e(o.b,[e("img",{staticStyle:{"max-height":"300px"},attrs:{alt:"methods",src:n(353)}}),t._v(" "),e("div",{staticClass:"mt-2"},[t._v("\n            • Even with removing half of the genome, there were only 226 genome that changed species."),e("br"),t._v("\n            • 184 of the genome changes were in nomenclaturely unimportant species (i.e. placeholder species)."),e("br"),t._v("\n            • The 226 genomes belonged to 206 unique species clusters from 134 unique genera.\n          ")])])],1)],1)],1),t._v(" "),e(c.a,{staticClass:"mt-3"},[e(o.c,{staticClass:"justify-center text-white bg-purple"},[t._v("\n      The perfect storm for species reclassification\n    ")]),t._v(" "),e(o.b,{staticClass:"text-center pt-4"},[e(c.a,{attrs:{outlined:""}},[e(o.b,[e("img",{staticStyle:{"max-height":"400px"},attrs:{alt:"methods",src:n(354)}}),t._v(" "),e("div",{staticClass:"mt-2"},[t._v("\n            • Chimeric assemblies decrease interspecies ANI to closely related species."),e("br"),t._v("\n            • Removal of foreign contigs increases ANI to neighbouring species clusters."),e("br"),t._v("\n            • If interspecies ANI similarity is extremely close, a species change is more likely."),e("br"),t._v("\n            • Placeholder species are disproportionately affected.\n          ")])])],1),t._v(" "),e(c.a,{staticClass:"mt-4",attrs:{outlined:""}},[e(o.c,{staticClass:"justify-center"},[t._v("\n          Whole genome alignment shows C. braakii contamination\n        ")]),t._v(" "),e(o.b,[e("img",{staticStyle:{"max-height":"400px"},attrs:{alt:"methods",src:n(355)}})])],1),t._v(" "),e(c.a,{staticClass:"mt-4",attrs:{outlined:""}},[e(o.c,{staticClass:"justify-center"},[t._v("\n          Removing contamination merges K. sp900759445 and K. planticola\n        ")]),t._v(" "),e(o.b,[e("img",{staticStyle:{"max-height":"400px"},attrs:{alt:"methods",src:n(356)}})])],1)],1)],1),t._v(" "),e(c.a,{staticClass:"mt-3"},[e(o.c,{staticClass:"justify-center text-white bg-purple"},[t._v("\n      Closely related genomes blur interspecies boundaries\n    ")]),t._v(" "),e(o.b,{staticClass:"text-center pt-4"},[e(c.a,{attrs:{outlined:""}},[e(o.b,[e("div",{staticClass:"mt-2"},[t._v("\n            • 76.75% of the interspecies ANI comparisons were >92% ANI for the 226 genomes that changed species,\n            compared to 16.58% for all genomes in GTDB R207."),e("br"),t._v("\n            • Significantly more non-representative genomes fall within multiple species ANI radii."),e("br"),t._v("\n            • Genomes are more likely to have a species reassignment if they are close to multiple representatives.\n          ")])])],1),t._v(" "),e(c.a,{staticClass:"mt-4",attrs:{outlined:""}},[e(o.c,{staticClass:"justify-center"},[t._v("\n          Species assignments can be on a knifes edge\n        ")]),t._v(" "),e(o.b,[e("img",{staticStyle:{"max-height":"400px"},attrs:{alt:"methods",src:n(357)}})])],1),t._v(" "),e(c.a,{staticClass:"mt-4",attrs:{outlined:""}},[e(o.c,{staticClass:"justify-center"},[t._v("\n          Interspecies ANI is higher for genomes that changed species\n        ")]),t._v(" "),e(o.b,[e("img",{staticStyle:{"max-height":"400px"},attrs:{alt:"methods",src:n(358)}})])],1)],1)],1),t._v(" "),e(c.a,{staticClass:"mt-3",attrs:{color:"#4f2984"}},[e(o.b,{staticClass:"text-center pt-4"},[e(l.a,{attrs:{"no-gutters":""}},[e(r.a,[e("img",{attrs:{alt:"methods",src:n(311),height:"75"}})]),t._v(" "),e(r.a,[e("img",{attrs:{alt:"methods",src:n(312),height:"50"}})]),t._v(" "),e(r.a,[e("img",{attrs:{alt:"methods",src:n(313),height:"50"}})])],1)],1)],1)],1)}),[],!1,null,"0b2808f0",null);e.default=component.exports}}]);